master/pkg/schemas/merge.go

```diff
@@ -0,0 +1,187 @@
+	if obj.Kind() != reflect.Ptr {
+		panic("non-pointer in merge")
+	}
+	// obj can't be a nil pointer, because Merge(nil, y) doesn't make any sense.
+	if obj.IsZero() {
+		panic("nil pointer in merge")
+	}
+	// Now operate on what obj points to.
+	obj = obj.Elem()
+
+	switch obj.Kind() {
+	case reflect.Interface:
+		if obj.IsZero() {
+			// This doesn't make any sense; we need a type.
+			panic("got a nil interface as the obj to merge into")
+		}
+		// Dereference the type but not the original pointer.
+		merge(obj.Elem().Addr(), src, allocated, name)
+
+	case reflect.Ptr:
+		// Note that this is a double-pointer since we already dereference the object once.
+		if obj.IsZero() {
+			// Allocate, then recurse with allocated = true.
+			tmp := reflect.New(obj.Type().Elem())
+			obj.Set(tmp)
+			merge(obj, src, true, name)
+		} else {
+			merge(obj, src, false, name)
+		}
+
+	case reflect.Struct:
+		assertTypeMatch(obj, src)
+		// Detect what to do with union fields.  There are 4 important cases:
+		//  1. src has a union member, obj does not -> recurse into that field.
+		//  2. src has a union member, obj has the same one -> recurse into that field.
+		//  3. src has a union member, obj has the different one -> do not recurse.
+		//  4. src has no union member -> recursing is a noop and doesn't matter
+		// Logically, this reduces to:
+		//   - if obj has a union member, src does not have the same one -> don't recurse.
+		//   - else -> recurse
+		recurseIntoUnion := true
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok {
+				if !obj.Field(i).IsZero() && src.Field(i).IsZero() {
+					recurseIntoUnion = false
+					break
+				}
+			}
+		}
+		// Recurse into each field of the struct.
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			srcField := src.Field(i)
+			objField := obj.Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok && !recurseIntoUnion {
+				continue
+			}
+			fieldName := fmt.Sprintf("%v.%v", name, structField.Name)
+			merge(objField.Addr(), srcField, false, fieldName)
+		}
+
+	case reflect.Map:
+		assertTypeMatch(obj, src)
+		// Maps get fused together; all input keys are written into the output map.
+		for _, key := range src.MapKeys() {
+			// Ensure key is not already set in obj.
+			if objVal := obj.MapIndex(key); objVal.IsValid() {
+				continue
+			}
+			elemName := fmt.Sprintf("%v.[%v]", name, key.Interface())
+			val := src.MapIndex(key)
+
+			cpy := reflect.New(val.Type())
+			merge(cpy, val, true, elemName)
+
+			// Update the original value with the defaulted value.
+			// Deref both layers of pointers we created.
+			obj.SetMapIndex(key, cpy.Elem())
+		}
+
+	case reflect.Slice:
+		assertTypeMatch(obj, src)
+		if !allocated {
+			// Don't override non-newly-allocated slices.
+			return
+		}
+		// Slices get copied only if the original was nil.
+		for i := 0; i < src.Len(); i++ {
```

Comment:
I think I commented this in the python, for some reason I thought we wanted to merge differently, but the go is so verbose i know it's on purpose now 😆